### DISCUSSION TITLE
Ending video trial at a random point in the video

### DISCUSSION THREAD

**Discussion by jackransomlovell** [jspsych/jsPsych#1101] [Q&A] [answered]
Hi all, I am currently trying to randomly select a stopping point for a video that is presented in an `HTML-keyboard-response` trial. My current approach is attached below, where I am to get the duration of the video then use `Math.random()` to get a random integer between 0 and the video duration and pass said integer into the `trial_duration:` parameter. I am currently stuck and don't have any idea how to move forward.  The array that is sampled called `video` and is a list of amazon s3 URLs where the videos for the experiment live. Thanks, Jack

```js
{
        type: 'html-keyboard-response',
        stimulus: function(){
            var vid = jsPsych.randomization.sampleWithoutReplacement(video, 1);
            var vid_tag = '<video id="stim" autoplay width="640" height="480">'+                                                          
                            '<source src='+vid+' type="video/mp4">'+ 
                            '<source src='+vid+' type="video/mov">'+        
                          '</video>'
            vid_stim.push(vid_tag);
            trial_stim.push(vid_tag);
            return(vid_tag)
        },
        choices: jsPsych.NO_KEYS,
        on_load:setTimeout(function(){
            var duration = document.getElementById("stim").duration
            var stop = Math.floor(Math.random() * duration)
            vid_duration.push(stop)},1000),
        trial_duration: vid_duration[vid_duration.length],
        data:{
            video: function(){
                var string = trial_stim[trial_stim.length]
                return string
            },
            phase: 'response'
        }
}
```

**Comment by becky-gilbert** [comment, marked as answer]
Hi Jack,

I can see what you were trying to do here but unfortunately this approach won't work. When you use [functions as parameters](https://www.jspsych.org/overview/trial/#dynamic-parameters) for 'normal' (non-event-related) parameters like `stimulus` and `trial_duration`, those functions will be evaluated immediately _before the trial starts_. But here you're using the trial's `on_load` function to get the stop time, and then trying to use that value inside of the `trial_duration` function. This won't work because the `on_load` function runs _after_ the trial starts (i.e. after the `trial_duration` function has already been called). 

Is there any way to get the video durations before each trial starts? For instance, could you add the durations manually for each video? Or maybe you could write some JavaScript that gets the video durations dynamically when the HTML page first loads?

A couple other things:
- The [jsPsych.randomization.sampleWithoutReplacement](https://www.jspsych.org/core_library/jspsych-randomization/#jspsychrandomizationsamplewithoutreplacement) function always returns an array, even when you're only sampling one element. So you need to use `[0]` to get the first (only) element inside the array that the function returns.
```javascript
var vid = jsPsych.randomization.sampleWithoutReplacement(video, 1)[0];
```
- FYI, the [video-keyboard-response plugin](https://www.jspsych.org/plugins/jspsych-video-keyboard-response/) includes a `stop` parameter that specifies when the video should stop. That might be useful if you end up specifying the video stop times manually and want to set that value separately from the trial duration.
- Remember that JavaScript indexing starts at 0, so if you want the last element in an array, you need to use array.length-1, rather than array.length (e.g. `trial_stim[trial_stim.length-1]`)
- I could be wrong, but in a trial's `data` parameter, I don't think it's possible to put a function inside of an object. If you need to use a variable in the trial data, then I think the value of `data` must be a function that returns an object, but you can use variables in the object, like this:
```javascript
data: function(){ 
  return {video: trial_stim[trial_stim.length-1], phase: 'response' }
}
```

**Comment by jackransomlovell** [comment]
Thanks for the help, Becky!

I have ended up just "hard-coding" each video's duration and using if and
if-else statements. This obviously has it's problems and is not the most
optimal way to code the experiment. I think once I find a bit more time I'd
like to write a function that is able to get the video's duration before
loading. I think you are right with the <data> parameter, and I have fixed
my code accordingly.

Thank you for your help!
